event and resource docs first pass: (#8858)
PrefectHQ/prefect: docs/concepts/events-and-resources.md

---

> description: Events are the primary units of observability in Prefect Cloud

Reviewer: ```suggestion
description: An event is a notification of a change from a resource
```

tags:
    - resources
    - events
    - observability
---

# Events

> Events are the primary unit of observability in Prefect Cloud, and are designed to be a simple but complete record of something occuring somewhere in your stack at a given time, and power several features in prefect cloud, including flow run logs, audit logs, and automations. Prefect events can represent API calls, state changes, or events from your execution environment or infrastructure. Events enable analysis of your data stack via [event feed](/ui/events/), and the confiuration of stack reactivity via [automations](/ui/automations/).

Reviewer: ```suggestion
An event is a notification of a change. Together, events form a feed of activity recording what's happening across your stack. Events power several features in Prefect Cloud, including flow run logs, audit logs, and automations. Events can represent API calls, state transitions, or changes in your execution environment or infrastructure. Events enable observability into your data stack via [event feed](/ui/events/), and the configuration of Prefect's reactivity via [automations](/ui/automations/).
```

Reviewer: FWIW, I don't think is accurate to describe events as a "complete" record. I would prefer to reserve the word "analysis" for more synthetic representations of the information.

> Events are the primary unit of observability in Prefect Cloud, and are designed to be a simple but complete record of something occuring somewhere in your stack at a given time, and power several features in prefect cloud, including flow run logs, audit logs, and automations. Prefect events can represent API calls, state changes, or events from your execution environment or infrastructure. Events enable analysis of your data stack via [event feed](/ui/events/), and the confiuration of stack reactivity via [automations](/ui/automations/).

Reviewer 2: - capitalization on `prefect cloud`
- "via the event feed" (added "the")
- spelling of `configuration`

![Prefect UI](../img/ui/event-feed.png)

## Event Specificiation

> Events adhere to a structured [specification](https://app.prefect.cloud/api/docs#tag/Events) for consistency.

Reviewer 2: i don't think you need to justify having a specification with "for consistency"

![Prefect UI](../img/ui/event-spec.png)
  
| Name | Type | Required? | Description |
| ---- | ---- | --------- | ----------- |

> | occurred |  String | yes | When the event happened from the sender's perspective |

Author: ```suggestion
| occurred |  String | yes | When the event happened |
```

> | occurred |  String | yes | When the event happened from the sender's perspective |

Reviewer: Would it be simpler to say that it is the time that the event was sent? Is that accurate?

| event |  String | yes | The name of the event that happened |
| resource|  Object | yes | The primary Resource this event concerns |
| related | Array | no | A list of additional Resources involved in this event |
| payload | Object | no | An open-ended set of data describing what happened |
| id | String | yes | The client-provided identifier of this event |
| follows | String | no | The ID of an event that is known to have occurred prior to this one. |

## Event Grammar

> Generally, events have a consistent and informative grammar - an event describes a resource and an action that a resource took or that was taken on that resource. For example, events emitted by Prefect objects take the form of:

Reviewer: ```suggestion
Generally, events have a consistent and informative grammar - an event describes a resource and an action that the resource took or that was taken on that resource. For example, events emitted by Prefect objects take the form of:
```

> Generally, events have a consistent and informative grammar - an event describes a resource and an action that a resource took. For example, events emitted by Prefect objects take the form of:

Reviewer: ```suggestion
Events have a consistent and informative grammar - an event describes a resource and an action that a resource took. For example, events emitted by Prefect objects take the form of:
```

```
prefect.block.write-method.called
prefect-cloud.automation.action.executed
prefect-cloud.user.logged-in
```

## Event Sources

> Events are automatically emitted by all prefect objects, including flows, tasks, deployments, work queues, and logs. Prefect-emitted events will contain the `prefect` or `prefect-cloud` resource prefix. Events can also be sent to the prefect [events API](https://app.prefect.cloud/api/docs#tag/Events) via authenticated http request.

Reviewer 2: prefect capitalization

> Prefect provides an events client that emits events, for use in arbitrary python code that may not be a task or 

Reviewer: ```suggestion
The Prefect SDK provides a method that emits events, for use in arbitrary python code that may not be a task or 
```

flow. Running the following code will emit events to Prefect Cloud, which will validate and ingest the event data.

```python3
from prefect.events import emit_event

def some_function(name: str="kiki") -> None:
    print(f"hi {name}!")
    emit_event(event=f"{name}.sent.event!", resource=f"prefect.resource.id":"coder.{name}")
          
some_function()
```

> This event now appears in the [event feed](/ui/events/) where you can visualize activity in context and configure [automations](/ui/automations/) to react to the presence or absence of it in the future.

Reviewer: ```suggestion
Emitted events will appear in the [event feed](/ui/events/) where you can visualize activity in context and configure [automations](/ui/automations/) to react to the presence or absence of it in the future.
```

## Resources

Every event has a primary resource, which describes what object emitted an event. Resources are used as quasi-stable identifiers for sources of events, and are consutrcted as dot-delimited strings such as:

```
prefect-cloud.automation.5b9c5c3d-6ca0-48d0-8331-79f4b65385b3.action.0
acme.user.kiki.elt_script_1
prefect.flow-run.e3755d32-cec5-42ca-9bcd-af236e308ba6
```

> Resources can optionally have additional arbitrary labels which can be used in different event aggregatino queries, such as:

Reviewer: ```suggestion
Resources can optionally have additional arbitrary labels which can be used in event aggregation queries, such as:
```

---
description: Events and resources are the primary units of observability in Prefect Cloud
tags:
    - resources
    - events
    - observability
---

# Events

> Events are the primary unit of observability in Prefect Cloud, and are designed to be a simple but complete record of something occuring somewhere in your stack at a given time. Prefect events can represent API calls, state changes, or events from your execution environment or infrastructure. Events enable both analysis of your data stack via (event feeds[link]) and the confiuration of stack reactivity via (automations[link])

Reviewer 3: "Events and resources are the primary units of observability..."
"Events are the primary unit of observability..."

Should these be consistent?
